src/audio/param.rs

```diff
@@ -0,0 +1,173 @@
+use audio::block::Tick;
+use audio::node::BlockInfo;
+
+/// An AudioParam. 
+///
+/// https://webaudio.github.io/web-audio-api/#AudioParam
+#[derive(Debug)]
+pub struct Param {
+    val: f32,
+    kind: ParamKind,
+    events: Vec<AutomationEvent>,
+    current_event: usize,
+    event_start_time: Tick,
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub enum ParamKind {
+    /// Value is held for entire block
+    KRate,
+    /// Value is updated each frame
+    ARate,
+}
+
+
+impl Param {
+    pub fn new(val: f32) -> Self{
+        Param {
+            val,
+            kind: ParamKind::ARate,
+            events: vec![],
+            current_event: 0,
+            event_start_time: Tick(0),
+        }
+    }
+
+    /// Update the value of this param to the next
+    ///
+    /// Returns true if anything changed
+    pub fn update(&mut self, block: &BlockInfo, tick: Tick) -> bool {
+        if tick.0 != 0 && self.kind == ParamKind::KRate {
+            return false;
+        }
+
+        // println!("Curr {:?}", self.current_event);
+        if self.events.len() <= self.current_event  {
+            return false;
+        }
+
+
+        let current_tick = block.absolute_tick(tick);
+        // println!("curr_tick {:?}", current_tick);
+        let mut current_event = &self.events[self.current_event];
+
+        // move to next event if necessary
+        // XXXManishearth k-rate events may get skipped over completely by this
+        // method. Firefox currently doesn't support these, however, so we can
+        // handle those later
+        if let Some(done_time) = current_event.done_time() {
+            if done_time < current_tick {
+                self.current_event += 1;
+                self.event_start_time = current_tick;
+                if let Some(next) = self.events.get(self.current_event) {
+                    current_event = next;
+                } else {
+                    return false;
+                }
+            }
+        } else if let Some(next) = self.events.get(self.current_event + 1) {
+            if let Some(start_time) = next.start_time() {
+                if start_time >= current_tick {
+                    self.current_event += 1;
+                    self.event_start_time = current_tick;
+                    current_event = next;
+                }
+            }
+        }
```

Comment:
Should we remove old events from `self.events`?

Reply:
Eventually yeah. I want to wait till I implement the rest of the events because some of them have  _some_ lookbehind and I'm not clear on how that goes, exactly.